File: Sources/DebridStreamer/Services/Debrid/PremiumizeService.swift
import Foundation

/// Premiumize API integration.
/// API docs: https://www.premiumize.me/api
actor PremiumizeService: DebridServiceProtocol {
    let serviceType = DebridServiceType.premiumize
    private let apiToken: String
    private let baseURL = "https://www.premiumize.me/api"
    private let session: URLSession

    init(apiToken: String, session: URLSession = .shared) {
        self.apiToken = apiToken
        self.session = session
    }

    func checkCache(hashes: [String]) async throws -> [String: CacheStatus] {
        guard !hashes.isEmpty else { return [:] }

        var results: [String: CacheStatus] = [:]
        let chunks = hashes.chunked(into: 100)

        for chunk in chunks {
            let itemsParam = chunk.map { "items[]=\($0)" }.joined(separator: "&")
            let data = try await requestRaw(path: "/cache/check", method: "GET", queryParams: itemsParam)

            if let json = try JSONSerialization.jsonObject(with: data) as? [String: Any],
               let response = json["response"] as? [Bool],
               let filenames = json["filename"] as? [String?],
               let filesizes = json["filesize"] as? [Int64?] {
                for (i, hash) in chunk.enumerated() where i < response.count {
                    let lowerHash = hash.lowercased()
                    if response[i] {
                        let name = i < filenames.count ? filenames[i] : nil
                        let size = i < filesizes.count ? filesizes[i] : nil
                        results[lowerHash] = .cached(fileId: nil, fileName: name, fileSize: size)
                    } else {
                        results[lowerHash] = .notCached
                    }
                }
            }
        }

        return results
    }

    func addMagnet(hash: String) async throws -> String {
        let magnet = "magnet:?xt=urn:btih:\(hash)"
        let body = "src=\(magnet.addingPercentEncoding(withAllowedCharacters: .urlQueryAllowed) ?? magnet)"
        let data = try await requestRaw(path: "/transfer/create", method: "POST", body: body)

        guard let json = try JSONSerialization.jsonObject(with: data) as? [String: Any],
              let id = json["id"] as? String else {
            throw DebridError.downloadFailed("Failed to add magnet to Premiumize")
        }
        return id
    }

    func selectFiles(torrentId: String, fileIds: [Int]) async throws {
        // Premiumize doesn't require file selection
    }

    func getStreamURL(torrentId: String) async throws -> StreamInfo {
        let encodedId = torrentId.addingPercentEncoding(withAllowedCharacters: .urlQueryAllowed) ?? torrentId
        let data = try await requestRaw(
            path: "/transfer/directdl",
            method: "POST",
            body: "src_id=\(encodedId)"
        )

        guard let json = try JSONSerialization.jsonObject(with: data) as? [String: Any],
              let content = json["content"] as? [[String: Any]] else {
            throw DebridError.noFilesAvailable
        }

        let candidates = content.compactMap { item -> DebridFileCandidate? in
            guard let link = item["link"] as? String else { return nil }
            let path = item["path"] as? String ?? "Unknown"
            let size = int64Value(item["size"]) ?? 0
            return DebridFileCandidate(link: link, fileName: path, sizeBytes: size)
        }
        guard let selected = DebridFileSelector.selectBest(from: candidates) else {
            throw DebridError.noFilesAvailable
        }

        let filename = URL(fileURLWithPath: selected.fileName).lastPathComponent
        let size = selected.sizeBytes

        return StreamInfo(
            streamURL: selected.link,
            quality: VideoQuality.parse(from: filename),
            codec: VideoCodec.parse(from: filename),
            audio: AudioFormat.parse(from: filename),
            source: SourceType.parse(from: filename),
            sizeBytes: size,
            fileName: filename,
            debridService: "PM"
        )
    }

    func unrestrict(link: String) async throws -> URL {
        // Premiumize doesn't have a separate unrestrict — directdl handles it
        guard let url = URL(string: link) else {
            throw DebridError.downloadFailed("Invalid link")
        }
        return url
    }

    func validateToken() async throws -> Bool {
        do {
            _ = try await getAccountInfo()
            return true
        } catch {
            return false
        }
    }

    func getAccountInfo() async throws -> DebridAccountInfo {
        let data = try await requestRaw(path: "/account/info", method: "GET")

        guard let json = try JSONSerialization.jsonObject(with: data) as? [String: Any] else {
            throw DebridError.invalidToken
        }

        let customerId = json["customer_id"] as? String ?? "Unknown"
        let premiumUntil = json["premium_until"] as? TimeInterval

        return DebridAccountInfo(
            username: customerId,
            email: nil,
            premiumExpiry: premiumUntil.map { Date(timeIntervalSince1970: $0) },
            isPremium: premiumUntil != nil
        )
    }

    // MARK: - HTTP

    private func requestRaw(path: String, method: String, queryParams: String? = nil, body: String? = nil) async throws -> Data {
        var urlStr = baseURL + path
        if let queryParams = queryParams, !queryParams.isEmpty {
            urlStr += "?\(queryParams)"
        }

        guard let url = URL(string: urlStr) else {
            throw DebridError.networkError("Invalid URL: \(path)")
        }

        var request = URLRequest(url: url)
        request.httpMethod = method
        request.setValue("Bearer \(apiToken)", forHTTPHeaderField: "Authorization")
        request.setValue(apiToken, forHTTPHeaderField: "X-API-Key")

        if let body = body {
            let authBodyComponent = "apikey=\(apiToken.addingPercentEncoding(withAllowedCharacters: .urlQueryAllowed) ?? apiToken)"
            let composedBody = body.isEmpty ? authBodyComponent : "\(body)&\(authBodyComponent)"
            request.httpBody = composedBody.data(using: .utf8)
            request.setValue("application/x-www-form-urlencoded", forHTTPHeaderField: "Content-Type")
        }

        let (data, response) = try await session.data(for: request)

        guard let httpResponse = response as? HTTPURLResponse else {
            throw DebridError.networkError("Invalid response")
        }

        guard (200...299).contains(httpResponse.statusCode) else {
            if httpResponse.statusCode == 401 {
                throw DebridError.invalidToken
            }
            throw DebridError.httpError(httpResponse.statusCode, String(data: data, encoding: .utf8) ?? "")
        }

        return data
    }

    private func int64Value(_ value: Any?) -> Int64? {
        if let number = value as? NSNumber {
            return number.int64Value
        }
        if let intValue = value as? Int {
            return Int64(intValue)
        }
        if let stringValue = value as? String {
            return Int64(stringValue)
        }
        return nil
    }
}

File: web/src/services/debrid/PremiumizeService.ts
// Port of Sources/DebridStreamer/Services/Debrid/PremiumizeService.swift.
//
// fetch-based Premiumize client. Mirrors the Swift actor: chunked /cache/check
// (parallel response/filename/filesize arrays), /transfer/create add, the
// /transfer/directdl poll that streams the link verbatim (no separate
// unrestrict), and /account/info decoding. Auth is a Bearer + X-API-Key header
// plus an `apikey=` form-body component (never in the query). The poll SLEEP is
// injectable (default no-op) so tests don't wait.

import {
  CacheStatus,
  type DebridAccountInfo,
  type DebridFileCandidate,
  DebridFileSelector,
  type DebridServiceType,
  DebridServiceType as DebridServiceTypeNS,
  type StreamInfo,
  AudioFormat,
  lastPathComponent,
  SourceType,
  VideoCodec,
  VideoQuality,
} from "./models";
import {
  type DebridService,
  DebridError,
  type FetchImpl,
  defaultFetchImpl,
  urlQueryEncode,
} from "./types";
import {
  asObjectArray,
  int64Value,
  noopSleep,
  parseJSONObject,
  type Sleep,
} from "./jsonHelpers";

function chunked<T>(arr: T[], size: number): T[][] {
  const out: T[][] = [];
  for (let i = 0; i < arr.length; i += size) out.push(arr.slice(i, i + size));
  return out;
}

export class PremiumizeService implements DebridService {
  readonly serviceType: DebridServiceType = DebridServiceTypeNS.premiumize;
  private readonly apiToken: string;
  private readonly baseURL = "https://www.premiumize.me/api";
  private readonly fetchImpl: FetchImpl;
  private readonly sleep: Sleep;

  constructor(apiToken: string, fetchImpl?: FetchImpl, sleep: Sleep = noopSleep) {
    this.apiToken = apiToken;
    this.fetchImpl = fetchImpl ?? defaultFetchImpl();
    this.sleep = sleep;
  }

  async checkCache(hashes: string[]): Promise<Record<string, CacheStatus>> {
    if (hashes.length === 0) return {};

    const results: Record<string, CacheStatus> = {};
    for (const chunk of chunked(hashes, 100)) {
      const itemsParam = chunk.map((h) => `items[]=${h}`).join("&");
      const data = await this.requestRaw("/cache/check", "GET", itemsParam);

      const json = parseJSONObject(data);
      const response = json && Array.isArray(json.response) ? json.response : null;
      const filenames = json && Array.isArray(json.filename) ? json.filename : null;
      const filesizes = json && Array.isArray(json.filesize) ? json.filesize : null;
      if (response && filenames && filesizes) {
        chunk.forEach((hash, i) => {
          if (i >= response.length) return;
          const lowerHash = hash.toLowerCase();
          if (response[i] === true) {
            const name =
              i < filenames.length && typeof filenames[i] === "string"
                ? (filenames[i] as string)
                : null;
            const size =
              i < filesizes.length && typeof filesizes[i] === "number"
                ? (filesizes[i] as number)
                : null;
            results[lowerHash] = CacheStatus.cached(null, name, size);
          } else {
            results[lowerHash] = CacheStatus.notCached;
          }
        });
      }
    }

    return results;
  }

  async addMagnet(hash: string): Promise<string> {
    const magnet = `magnet:?xt=urn:btih:${hash}`;
    const body = `src=${urlQueryEncode(magnet)}`;
    const data = await this.requestRaw("/transfer/create", "POST", undefined, body);

    const json = parseJSONObject(data);
    const id = json && typeof json.id === "string" ? json.id : null;
    if (id == null) {
      throw DebridError.downloadFailed("Failed to add magnet to Premiumize");
    }
    return id;
  }

  async selectFiles(_torrentId: string, _fileIds: number[]): Promise<void> {
    // Premiumize doesn't require file selection.
  }

  async getStreamURL(torrentId: string): Promise<StreamInfo> {
    const encodedId = urlQueryEncode(torrentId);

    const maxAttempts = 20;
    let content: Record<string, unknown>[] = [];

    for (let attempt = 0; attempt < maxAttempts; attempt++) {
      const data = await this.requestRaw(
        "/transfer/directdl",
        "POST",
        undefined,
        `src_id=${encodedId}`,
      );

      const json = parseJSONObject(data);
      const items = json && asObjectArray(json.content);
      if (items && items.length > 0) {
        content = items;
        break;
      }

      if (attempt < maxAttempts - 1) {
        await this.sleep(1000);
      }
    }

    if (content.length === 0) throw DebridError.noFilesAvailable();

    const candidates: DebridFileCandidate[] = [];
    for (const item of content) {
      if (typeof item.link !== "string") continue;
      const path = typeof item.path === "string" ? item.path : "Unknown";
      const size = int64Value(item.size) ?? 0;
      candidates.push({ link: item.link, fileName: path, sizeBytes: size });
    }
    const selected = DebridFileSelector.selectBest(candidates);
    if (selected == null) throw DebridError.noFilesAvailable();

    const filename = lastPathComponent(selected.fileName);
    const size = selected.sizeBytes;

    return {
      streamURL: selected.link,
      quality: VideoQuality.parse(filename),
      codec: VideoCodec.parse(filename),
      audio: AudioFormat.parse(filename),
      source: SourceType.parse(filename),
      sizeBytes: size,
      fileName: filename,
      debridService: "PM",
    };
  }

  async unrestrict(link: string): Promise<string> {
    // Premiumize doesn't have a separate unrestrict — directdl handles it.
    if (!isAbsoluteURL(link)) throw DebridError.downloadFailed("Invalid link");
    return link;
  }

  async validateToken(): Promise<boolean> {
    try {
      await this.getAccountInfo();
      return true;
    } catch {
      return false;
    }
  }

  async getAccountInfo(): Promise<DebridAccountInfo> {
    const data = await this.requestRaw("/account/info", "GET");
    const json = parseJSONObject(data);
    if (json == null) throw DebridError.invalidToken();

    const customerId = typeof json.customer_id === "string" ? json.customer_id : "Unknown";
    const premiumUntil =
      typeof json.premium_until === "number" ? json.premium_until : null;

    return {
      username: customerId,
      email: null,
      premiumExpiry: premiumUntil != null ? new Date(premiumUntil * 1000) : null,
      isPremium: premiumUntil != null,
    };
  }

  // MARK: - HTTP

  private async requestRaw(
    path: string,
    method: string,
    queryParams?: string,
    body?: string,
  ): Promise<string> {
    let urlStr = this.baseURL + path;
    if (queryParams != null && queryParams.length > 0) {
      urlStr += `?${queryParams}`;
    }

    const headers: Record<string, string> = {
      Authorization: `Bearer ${this.apiToken}`,
      "X-API-Key": this.apiToken,
    };

    let composedBody: string | undefined;
    if (body != null) {
      const authBodyComponent = `apikey=${urlQueryEncode(this.apiToken)}`;
      composedBody = body.length === 0 ? authBodyComponent : `${body}&${authBodyComponent}`;
      headers["Content-Type"] = "application/x-www-form-urlencoded";
    }

    const response = await this.fetchImpl(urlStr, {
      method,
      headers,
      body: composedBody,
    });
    const status = response.status;

    if (!(status >= 200 && status <= 299)) {
      if (status === 401) throw DebridError.invalidToken();
      const errText = (await response.text().catch(() => "")) || "";
      throw DebridError.httpError(status, errText);
    }

    return response.text();
  }
}

function isAbsoluteURL(value: string): boolean {
  try {
    new URL(value);
    return true;
  } catch {
    return false;
  }
}
